File: e2e/utils/api-helper.js
import { request } from '@playwright/test';

/**
 * Creates a unique test user via API for isolated E2E testing.
 * @param {string} baseURL The base URL of the backend API
 * @returns {Promise<Object>} The user credentials and data
 */
export async function createTestUser(baseURL = 'http://127.0.0.1:5001/api') {
  const uniqueSuffix = `${Date.now()}-${Math.random().toString(36).slice(2, 8)}`;
  const testUser = {
    fullName: 'Test Automation User',
    email: `auto.user+${uniqueSuffix}@example.com`,
    password: 'Password123!',
    phone: '1234567890'
  };

  const context = await request.newContext();
  const response = await context.post(`${baseURL}/auth/signup`, {
    data: testUser,
  });

  if (!response.ok()) {
    const text = await response.text();
    throw new Error(`Failed to create test user: ${response.status()} ${text}`);
  }

  const { token, user } = await response.json();
  
  return {
    credentials: testUser,
    token,
    user
  };
}

export async function createTransaction(token, payload, baseURL = 'http://127.0.0.1:5001/api') {
  const context = await request.newContext();
  const response = await context.post(`${baseURL}/transactions`, {
    headers: { Authorization: `Bearer ${token}` },
    data: payload,
  });

  if (!response.ok()) {
    const text = await response.text();
    throw new Error(`Failed to create transaction: ${response.status()} ${text}`);
  }

  return response.json();
}

File: selenium/src/main/java/utilities/ApiHelper.java
package utilities;

import io.restassured.RestAssured;
import io.restassured.http.ContentType;
import io.restassured.response.Response;
import net.datafaker.Faker;
import org.json.JSONObject;

public class ApiHelper {
    private static final String BASE_URL = "http://127.0.0.1:5001/api";
    private static final Faker faker = new Faker();

    public static JSONObject createTestUser() {
        String uniqueSuffix = String.valueOf(System.currentTimeMillis());
        String email = "auto.user+" + uniqueSuffix + "@example.com";
        String password = "Password123!";
        
        JSONObject requestParams = new JSONObject();
        requestParams.put("fullName", faker.name().fullName());
        requestParams.put("email", email);
        requestParams.put("password", password);
        requestParams.put("phone", faker.phoneNumber().cellPhone());

        Response response = RestAssured.given()
                .contentType(ContentType.JSON)
                .body(requestParams.toString())
                .post(BASE_URL + "/auth/signup");

        if (response.getStatusCode() != 200 && response.getStatusCode() != 201) {
            throw new RuntimeException("Failed to create test user: " + response.getBody().asString());
        }

        JSONObject result = new JSONObject();
        result.put("email", email);
        result.put("password", password);
        result.put("response", new JSONObject(response.getBody().asString()));

        return result;
    }
}
